File: typescript/src/listMarkets.ts
import { getMarkets } from "./client";

interface Market {
  name: string;
  index: number;
  szDecimals: number;
}

async function main() {
  const data: { perps: Market[]; spot: Market[]; hip3: Record<string, Market[]> } = await getMarkets();

  const { perps, spot, hip3 } = data;

  const hip3All: (Market & { dex: string })[] = [];
  for (const [dex, markets] of Object.entries(hip3)) {
    for (const m of markets) {
      hip3All.push({ ...m, dex });
    }
  }

  console.log(`Perps: ${perps.length}  |  Spot: ${spot.length}  |  HIP-3: ${hip3All.length}\n`);

  for (const [label, group] of [["Perps", perps], ["Spot", spot]] as [string, Market[]][]) {
    if (group.length > 0) {
      console.log(`--- ${label} ---`);
      for (const m of group.slice(0, 10)) {
        console.log(`  ${m.name.padEnd(16)}  index=${m.index}  szDecimals=${m.szDecimals}`);
      }
      if (group.length > 10) {
        console.log(`  ... and ${group.length - 10} more`);
      }
      console.log();
    }
  }

  if (hip3All.length > 0) {
    console.log("--- HIP-3 ---");
    for (const m of hip3All.slice(0, 10)) {
      const display = `${m.dex}:${m.name}`;
      console.log(`  ${display.padEnd(16)}  index=${m.index}  szDecimals=${m.szDecimals}`);
    }
    if (hip3All.length > 10) {
      console.log(`  ... and ${hip3All.length - 10} more`);
    }
    console.log();
  }
}

main();

File: python/list_markets.py
"""List all available markets (perps, spot, HIP-3)."""

from client import get_markets

data = get_markets()

perps = data.get("perps", [])
spot = data.get("spot", [])
hip3_raw = data.get("hip3", {})

hip3_all = []
for dex, markets in hip3_raw.items():
    for m in markets:
        hip3_all.append({**m, "dex": dex})

print(f"Perps: {len(perps)}  |  Spot: {len(spot)}  |  HIP-3: {len(hip3_all)}\n")

for label, group in [("Perps", perps), ("Spot", spot)]:
    if group:
        print(f"--- {label} ---")
        for m in group[:10]:
            print(f"  {m['name']:16s}  index={m['index']}  szDecimals={m['szDecimals']}")
        if len(group) > 10:
            print(f"  ... and {len(group) - 10} more")
        print()

if hip3_all:
    print("--- HIP-3 ---")
    for m in hip3_all[:10]:
        display = f"{m['dex']}:{m['name']}"
        print(f"  {display:16s}  index={m['index']}  szDecimals={m['szDecimals']}")
    if len(hip3_all) > 10:
        print(f"  ... and {len(hip3_all) - 10} more")
    print()
